SOAP with history

boxlady opened the pull request:
Fixes #

#### Short description of what this resolves:
This adds history to the soap. Which shows a list of all the previous soap entries for this patient above the entry textbox.

#### Changes proposed in this pull request:
Added entry history to the SOAP form


below added by bradymiller:
Up For Grabs demo for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Gamma_-_Up_For_Grabs_Demo

stephenwaite commented on interface/forms/soap/FormSOAP.class.php:
```diff
@@ -64,117 +69,190 @@ function __construct($id = "", $_prefix = "")
         //$this->temp_methods = parent::_load_enum("temp_locations",false);
     }
 
+    function histroy()
```
hi @boxlady :), simple typo here

stephenwaite commented on interface/forms/soap/templates/general_new.html:
```diff
@@ -1,68 +1,72 @@
-{headerTemplate}
+    <title>{xlt t='SOAP'}</title>
+    {headerTemplate}
 </head>
```
maybe title with `SOAP history` too?

stephenwaite commented on interface/forms/soap/FormSOAP.class.php:
```diff
@@ -35,15 +35,20 @@ class FormSOAP extends ORDataObject
      */
```
hi @boxlady , may want to create new with `interface/forms/soap_history/FormSOAPHistory.class.php` so preserve legendary SOAP :)

bradymiller commented:
hi @boxlady and @stephenwaite ,
I haven't tested this, but like the idea of having history available from prior forms. This can probably be worked in a way where a new form does not need to be created (since it is just showing prior entries). Maybe a tooltip type graphic that expands to show the history when clicked on. Will plan to test this out soon. -brady

bradymiller commented:
Up For Grabs demo for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Gamma_-_Up_For_Grabs_Demo

sunsetsystems commented:
Nice. For what it's worth LBF forms also have the ability to show past instances; see Repeats in the layout properties.

stephenwaite commented:
looks like when you edit it doubles up the saved history

boxlady commented:
> looks like when you edit it doubles up the saved history

You're right, if you click edit from soap box, it shows double, however if you open  it though the menu not.. I'll have to look into that.

boxlady commented:
I updated this, removed doubles, and ID's. It has been working in production for the last couple months. Is this something that can be added to master?

bradymiller commented:
hi @boxlady ,
Just tested this. Many users will not use this, however, it is a really neat thing to have and am guessing a significant amount of users will find it useful. So it doesn't detract for the users that don't use it, seems like would be best to collapse the history parts of the forms, which the user can then click on to expand them.
@stu01509 , since you are turning into the js/bootstrap expert, any thoughts on how to best do this?

stu01509 commented:
Hello 

Thanks for your mention, How can I test and run this PR on my environment?

stephenwaite commented:
hi @stu01509, for the easy dev enviro go to your local repo
0. docker-compose down -v
1. git remote add boxlady https://github.com/boxlady/openemr
2. git fetch boxlady
3. git checkout boxlady/soep_history
4. docker pull openemr/openemr:flex
5. docker-compose up -d
6. have a cup of :coffee: 
7. docker logs <openemr container id>
should show this after you've had a sip or two
![Screenshot from 2020-06-05 16-34-40](https://user-images.githubusercontent.com/5618372/83920554-77f22d80-a74a-11ea-8f39-3c30b5f22597.png)
8. go to http://localhost:8300/interface/login/login.php?site=default
9. create an encounter and choose a clinical->soap form

stu01509 commented:
Hi @boxlady  @bradymiller @stephenwaite  and @tywrenn 

I'm so sorry about that, I overlook this issue mention. Now I have deployed this version on my environment and try to use the accordion design the sample, What do you think, let me know your thought.

![image](https://user-images.githubusercontent.com/22230889/85934806-d3af7100-b91a-11ea-907b-bcc209202775.png)

bradymiller commented:
hi @stu01509 , I would think of terms of potentially 10-100 possible prior notes. So, to be least invasive, would probably just have a panel that says something like; click here to see prior <number message>, which would then expand if the user clicks on it. That way, it won't take up much space when not used since many users likely will just ignore it.

tywrenn commented:
@bradymiller @stu01509 For this situation I like accordions yet with the possibility of 10-100 of these seems like a lot. For each of the accordion btn's you should add the following classes: `btn-sm p-0`

This will change the button size to small and remove the padding so the accordion appears smaller. The output looks something like this:
![image](https://user-images.githubusercontent.com/46497068/85953603-a2956780-b93f-11ea-8cdd-a53c9030e163.png)

bradymiller commented:
Not sure how accordion acts. should:
1. place all the prior entries under one accordion (the provider can then see them all in one glance to quickly copy what text they want to).
2. don't lose visibility of the text field for quick pasting

stu01509 commented:
I guess it may look like this, Only click one accordion it will show all of the history messages.

![image](https://user-images.githubusercontent.com/22230889/85961643-7250d600-b9de-11ea-9d9d-96e6368558de.png)

bradymiller commented:
@stu01509 , looks great!

stu01509 commented:
Hi @bradymiller 

Happy to hear that, so, how do I create this feature?

bradymiller commented:
bump

bradymiller commented:
bump
